picard/tagger.py

```diff
@@ -466,6 +476,10 @@ def move_files(self, files, target):
         if target is None:
             log.debug("Aborting move since target is invalid")
             return
+
+        # Temporarily disable left and right panel sorting
+        self.window.panel.set_sorting(False)
```

Comment:
I was wondering about those, they are good candidates for self-made with statement.

I was thinking about be able to write `with SortingDisabled(self.window.panel):`, here is a small test script implementing the idea:

```python
class Obj:
    def __init__(self):
        self.sorting = True

    def set_sorting(self, value):
        self.sorting = value

    def __str__(self):
        return "Sorting: %r" % self.sorting


class SortingDisabled():
    def __init__(self, obj):
        self.obj = obj

    def __enter__(self):
        self.saved_value = self.obj.sorting
        self.obj.set_sorting(False)
        return self.obj

    def __exit__(self, type, value, traceback):
        self.obj.set_sorting(self.saved_value)
        return False

print("basic with statement")
obj = Obj()
print(obj)
with SortingDisabled(obj):
    print(obj)
print(obj)

print("with exception in with statement")
print(obj)
try:
    with SortingDisabled(obj):
        print(obj)
        raise Exception("Exception test raised")
except Exception as e:
    print(e)

print(obj)

print("with original value being False")
obj = Obj()
obj.sorting = False
print(obj)
with SortingDisabled(obj):
    print(obj)
print(obj)
```

What do you think? The idea behind is to ensure sorting value is restored whatever happens in between.

Reply:
Not really sure on this one. On the current implementation, due to concurrency, the worst thing that can happen is the sorting be enabled during the operation and things get slower (T1 T2 T1). On this one, the worst thing that can happen is the sorting to get disabled and not be re-enabled in the end or future calls. (T1 T2 T1 T2)

Reply:
If the other thread finishes before the first (e.g. move stuff while clustering is running).
It can be handled with either a semaphore or a lock+counter combo inside set_sorting. I assumed it wouldn't be a problem due to how fast things are, where there is no reason not to wait stuff finish processing before moving/editing/etc. 

I didn't touch on set_sorting during add_files and lookup/scanning for that exact reason, as they're bounded by IO and API, being far slower and having way higher probability of manual intervention (moving).

Reply:
How about the following?
```
        # Add 1 if sorting is disabled, subtract 1 if sorting is enabled
        with self._sorting_lock:
            self._sorting_count += -1 if sort else 1
            
        # Guarantee counter is >= 0
        self._sorting_count = 0 if self._sorting_count < 0 else self._sorting_count
        
        # If counter == 0, re-enable sorting
        sort = True if self._sorting_count == 0 else False    
        self.views[Panels.CLUSTERS.value].setSortingEnabled(sort)
        self.views[Panels.ALBUMS.value].setSortingEnabled(sort)
```

Should solve all cases. Once sorting is disabled, will remain disabled until all DisableEnable pairs have been called.

Reply:
OK, basically we switch from a boolean to an integer counter, each time a thread disables sorting it stays disabled until the last thread actually re-enables it.

I can't say I'm very enthusiastic, not because of the solution (it should work), but because of the problem.

Enabling/disabling sorting looks messy from start, and it becomes even uglier when concurrency comes into play.

@phw what do you think about this? Can we get rid of sorting on/off or should we improve current stuff so it plays well with concurrency?